src/app/ReadHandler.h

```diff
@@ -333,13 +332,22 @@ class ReadHandler : public Messaging::ExchangeDelegate
     bool IsIdle() const { return mState == HandlerState::Idle; }
 
     /// @brief Returns whether the ReadHandler is in a state where it can send a report and there is data to report.
-    bool IsReportable() const
+    bool ShouldStartReporting() const
+    {
+        // Important: Anything that changes ShouldStartReporting() from false to true
+        // (which can only happen for subscriptions) must call
+        // mObserver->OnBecameReportable(this).
+        return CanStartReporting() && (ShouldReportUnscheduled() || IsDirty());
+    }
+    /// @brief CanStartReporting() is true if the ReadHandler is in a state where it could generate
+    /// a (possibly empty) report if someone asked it to.
+    bool CanStartReporting() const { return mState == HandlerState::CanStartReporting; }
+    /// @brief ShouldReportUnscheduled() is true if the ReadHandler should be asked to generate reports
+    /// without consulting the report scheduler.
+    bool ShouldReportUnscheduled() const
     {
-        // Important: Anything that changes the state IsReportable must call mObserver->OnBecameReportable(this) for the scheduler
-        // to plan the next run accordingly.
-        return mState == HandlerState::GeneratingReports && IsDirty();
+        return CanStartReporting() && (IsType(ReadHandler::InteractionType::Read) || IsPriming()) && !IsActiveSubscription();
```

Comment:
I don't understand the IsActiveSubscription() test here.

Specifically:

1. For a Read-type handler, IsActiveSubscription is always false, so the test is always true.
2. For a subscribe-type handler, ActiveSubscription is never set until after PrimingReports is removed.  So if IsPriming() is true, IsActiveSubscription must be false.

So in either case, if `(IsType(ReadHandler::InteractionType::Read) || IsPriming())` is true, then `!IsActiveSubscription()` will also be true, right?

Reply:
As mentioned in another resolved conversation, we have one `MoveToState()` at the end of `OnStatusResponse()` that would end up skipping the scheduler because at that point the IsPriming() flag hasn't been cleared (it is cleared at the end of `SendSubscribeResponse()`) So we would end up sending an extra report there.